### PR TITLE
Refactored Attachment handling

### PR DESCRIPTION
**Pull request by catdawg**

### CODE REVIEW

**Comment by catdawg** [comment]
- Added a calculate layout method that takes a already built attributed string.
- Moved LayoutBehaviour to the types' OverflowPolicy but kept backwards compatibility.
- Attachments are no longer included in the span, and are handled as a special kind of character.
- Attachments are not cached by vector x anymore, they only request images with a target size, and it's up to the loadImage callback implementer to render accordingly.
- Attachments are rendered with the size of the current font instead of the size referenced from the json file.

**Comment by catdawg** [comment]
I fixed the conflicts here. Seems to be working on our side. Would be nice to get this merged, but I understand there are still some philosophical debates to be worked out around this change...

**Comment by dejan-gd** [comment]
I'll start working on this now.

File: vectorx/font/TextLine.hx [hunk 1]
```diff
@@ -282,7 +290,7 @@ class TextLine
                 lastSpanInLine.updateString();
             }
 
-            if(lastSpanInLine.haveShadow() && lastSpanInLine.attachmentId == null)
+            if(lastSpanInLine.haveShadow() /* && lastSpanInLine.attachmentId == null*/)
```
Comment (track3r) [inline vectorx/font/TextLine.hx:293]: Now calculation of line width would be broken in case there is an attachment at the end of a line.

Comment (catdawg) [inline vectorx/font/TextLine.hx:293]: damn I forgot about that comment. actually I think the shadows are not even working for attachments. So it does actually work for the line width, just not for the shadows...

**Comment by track3r** [comment]
I still don't see a point in this change "Attachments are no longer included in the span, and are handled as a special kind of character." You need it to edit internal string representation which I never intended to be editable. I don't see why you don't want to change templates in styled strings and then render them normally.

**Comment by catdawg** [comment]
what do you mean with change templates?

**Comment by track3r** [comment]
[f=arial22]%number_of_coins%{coint}[/f] -> [f=arial22]1 000 000{coint}[/f]

**Comment by catdawg** [comment]
The thing is that the attachments are visible and are kind of a character. So when I want to edit a section of the text, I don't want to care about the formatting by code, but I care about the attachments. I want to be able to swap for example %number_of_coins% with "no coins" or "10 {coin} !!!". Right now it's extremely easy to do this, but if the attachments spawn a new section, than we have no choice but to parse the original raw string and not use the already built AttributedString object to edit sections.

Also, an attachment is not formatting, so it's very weird that it is parsed as such, even if it makes the code easier to write, it doesn't make it easier to understand at all.

**Comment by dejan-gd** [comment]
Hi all,

I tried to merge this but it doesn't work as there are changes in at least buildunitylayout library which are not provided as a PR.

If I understood our collaboration concept well - we will need all changes in SGF libraries to be visible for us.

@catdawg please let me know your thoughts.

**Comment by jcxavier** [comment]
Hi @dejan-gd , since duellbuildunitylayout is not a public repo we cannot actually PR. But we can provide you with the patches.

**Comment by nensanders** [comment]
You can pull request on private repos.

**Comment by track3r** [comment]
> we have no choice but to parse the original raw string and not use the already built AttributedString object to edit sections

Honestly, I don't see why it's bad. Cost of constructing AttributedString from template is insignificant compared to layouting + rasterization work that is needed in any case

**Comment by dejan-gd** [comment]
Hi @catdawg, @jcxavier,
can't build it with the source from the ZIP. I'm getting this ouput:

```
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:25: characters 21-35 : game_engine.components.text.VectorTextComponent has no field text
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:27: characters 15-44 : game_engine.components.text.VectorTextComponent has no field horizontalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:40: characters 15-42 : game_engine.components.text.VectorTextComponent has no field verticalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:53: characters 15-40 : game_engine.components.text.VectorTextComponent has no field layoutBehaviour
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:70: characters 8-26 : game_engine.components.text.VectorTextComponent has no field text
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:75: characters 16-49 : game_engine.components.text.VectorTextComponent has no field horizontalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:78: characters 16-49 : game_engine.components.text.VectorTextComponent has no field horizontalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:81: characters 16-49 : game_engine.components.text.VectorTextComponent has no field horizontalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:87: characters 16-47 : game_engine.components.text.VectorTextComponent has no field verticalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:90: characters 16-47 : game_engine.components.text.VectorTextComponent has no field verticalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:93: characters 16-47 : game_engine.components.text.VectorTextComponent has no field verticalAlignment
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:99: characters 16-45 : game_engine.components.text.VectorTextComponent has no field layoutBehaviour
.duell/lib/game_engine/game_engine/serialization/serializers/VectorTextComponentSerializer.hx:102: characters 16-45 : game_engine.components.text.VectorTextComponent has no field layoutBehaviour
```

Please, could you provide complete patch for the changes in libraries game_engine and duellbuildunitylayout?

**Comment by dejan-gd** [comment]
As suggested by @catdawg - we will close this PR.

**Pull request closed without merging**